File: raytrace/src/intersections.rs
use crate::equals::EPSILON;
use crate::rays::Ray;
use crate::tuple::Tuple;
use crate::world::World;

#[derive(Debug,Copy,Clone)]
pub struct Intersection {
    pub t: f64,
    pub object: usize,
}

impl Intersection {
    pub fn equals( &self, other: Intersection ) -> bool {
        self.t == other.t && 
        self.object == other.object
    }
}

pub fn intersection(t: f64, object: usize) -> Intersection {
    Intersection { t, object }
}

pub fn intersections(args: &[Intersection]) -> Vec<Intersection> {
    let mut v = vec!();
    for arg in args {
        v.push(*arg);
    }
    v
}

pub fn hit( xs: Vec<Intersection> ) -> Option<Intersection> {
    let mut lowest = xs[0];
    for i in xs {
        if (lowest.t < 0.0 || i.t < lowest.t ) && i.t >= 0.0 {
            lowest = i;
        }
    }
    if lowest.t < 0.0 {
        None
    } else {
        Some( lowest )
    }
}

pub fn schlick(comps: Computations) -> f64 {
    let mut cos = comps.eyev.dot(&comps.normalv);

    // total internal reflection case
    if comps.n1 > comps.n2 {
        let n = comps.n1 / comps.n2;
        let sin2_t = n.powf(2.0) * (1.0 - cos.powf(2.0));
        if sin2_t > 1.0 {
            return 1.0 
        }

        cos = (1.0 - sin2_t).sqrt();
    }

    let r0 = ((comps.n1 - comps.n2) / (comps.n1 + comps.n2)).powf(2.0);
    r0 + (1.0 - r0) * (1.0 - cos).powf(5.0)
}

#[derive(Debug)]
pub struct Computations {
    pub t: f64,
    pub object: usize,
    pub point: Tuple,
    pub over_point: Tuple,
    pub under_point: Tuple,
    pub eyev: Tuple,
    pub normalv: Tuple,
    pub inside: bool,
    pub reflectv: Tuple,
    pub count: usize,
    pub n1: f64,   // refractive index coming FROM
    pub n2: f64,   // refractive index going TO
}

pub fn prepare_computations( hit: Intersection, 
                             r: Ray, 
                             w: &World, 
                             xs: &Vec<Intersection> ) -> Computations {
    let mut ins = false;
    let mut n = w.get_object(hit.object).normal_at( r.position(hit.t) );
    if n.dot( &-r.direction ) < 0.0 {
        n = -n;
        ins = true;
    }
    let op = r.position(hit.t) + n * EPSILON;
    let up = r.position(hit.t) - n * EPSILON;
    let rv = r.direction.reflect(&n);

    let mut containers: Vec<usize> = vec!();
    let mut n1 = 1.0;
    let mut n2 = 1.0;

    for i in xs {
        if i.equals(hit) {
            if containers.len() == 0 {
                n1 = 1.0;
            } else {
                let last = containers[containers.len()-1];
                n1 = w.get_object(last).get_material().refractive_index;
            }
        }

        if containers.contains(&i.object) {
            containers.retain(|x| *x != i.object);
        } else {
            containers.push(i.object);
        }

        if i.equals(hit) {
            if containers.len() == 0 {
                n2 = 1.0;
            } else {
                let last = containers[containers.len()-1];
                n2 = w.get_object(last).get_material().refractive_index;
            }
            break;
        }
    }

    Computations { 
        t: hit.t,
        object: hit.object,
        point: r.position( hit.t ),
        over_point: op,
        under_point: up,
        eyev: -r.direction,
        normalv: n,
        inside: ins,
        reflectv: rv,
        count: r.count,
        n1: n1,
        n2: n2,
    }
}

#[cfg(test)]
mod tests {
    use crate::intersections::{intersection, intersections, 
                               hit, prepare_computations, schlick};
    use crate::tuple::{point, vector};
    use crate::rays::ray;
    use crate::equals::equals;
    use crate::world::{world, default_world};
    use crate::planes::plane;
    use crate::spheres::glass_sphere;
    use crate::shapes::Shape;
    use crate::transform::{scaling, translation};
    use crate::materials::material;
    use crate::equals::EPSILON;
    use std::f64::consts::SQRT_2;

    #[test]
    fn intersection_creates_intersections(){
        let i = intersection(3.5, 1);

        assert_eq!( 3.5, i.t );
        assert_eq!( i.object, 1);
    }

    #[test]
    fn aggregating_intersections(){
        let i1 = intersection(1.0, 1);
        let i2 = intersection(2.0, 2);

        let xs = intersections(&[i1, i2]);
        assert_eq!( xs.len(), 2 );
        assert_eq!( xs[0].t, 1.0 );
        assert_eq!( xs[1].t, 2.0 );
    }

    #[test]
    fn hit_with_all_positive_intersections(){
        let i1 = intersection(1.0, 1);
        let i2 = intersection(2.0, 2);

        let xs = intersections(&[i2, i1]);
        let i = hit(xs);

        assert!( i.expect("positive intersections available").equals( i1 ));
    }

    #[test]
    fn hit_when_some_intersections_are_negative(){
        let i1 = intersection(-1.0, 1);
        let i2 = intersection(1.0, 2);

        let xs = intersections(&[i1, i2]);
        let i = hit(xs);

        assert!( i.expect("positive intersection available").equals( i2 ));
    }

    #[test]
    fn hit_when_all_intersections_are_negative(){
        let i1 = intersection(-2.0, 1);
        let i2 = intersection(-1.0, 2);

        let xs = intersections(&[i2, i1]);
        let i = hit(xs);

        assert!( match i {
                   Some(_x) => false,
                   None => true }
        );
    }

    #[test]
    fn hit_always_lowest_nonnegative_intersection(){
        let i1 = intersection( 5.0, 1);
        let i2 = intersection( 7.0, 2);
        let i3 = intersection(-3.0, 3);
        let i4 = intersection( 2.0, 4);

        let xs = intersections(&[i1, i2, i3, i4]);
        let i = hit(xs);

        assert!( i.expect("positive intersection available").equals( i4 ));
    }

    #[test]
    fn precomputing_intersection_state(){
        let r = ray( point(0.0, 0.0, -5.0), vector(0.0, 0.0, 1.0), 0 );
        let i = intersection(4.0, 1);
        let xs = intersections(&[i]);

        let comps = prepare_computations(i, r, &default_world(), &xs);

        assert!( equals(comps.t, i.t) );
        assert_eq!( comps.object, 1);
        assert!( comps.point.equals( point(0.0, 0.0, -1.0) ));
        assert!( comps.eyev.equals( vector(0.0, 0.0, -1.0) ));
        assert!( comps.normalv.equals( vector(0.0, 0.0, -1.0) ));
    }

    #[test]
    fn intersection_on_outside(){
        let r = ray( point(0.0, 0.0, -5.0), vector(0.0, 0.0, 1.0), 0 );
        let i = intersection(4.0, 1);
        let xs = intersections(&[i]);

        let comps = prepare_computations(i, r, &default_world(), &xs);

        assert!( comps.inside == false );
    }

    #[test]
    fn intersection_on_inside(){
        let r = ray( point(0.0, 0.0, 0.0), vector(0.0, 0.0, 1.0), 0 );
        let i = intersection(1.0, 1);
        let xs = intersections(&[i]);

        let comps = prepare_computations(i, r, &default_world(), &xs);

        assert!( comps.point.equals( point(0.0, 0.0, 1.0) ));
        assert!( comps.eyev.equals( vector(0.0, 0.0, -1.0) ));
        assert!( comps.normalv.equals( vector(0.0, 0.0, -1.0) ));
        assert!( comps.inside == true );
    }

    #[test]
    fn precomputing_reflection_vector(){
        let mut w = world();
        let p = plane();
        w.add_object(Box::new(p));
        let r = ray( point(0.0, 1.0, -1.0), vector(0.0, -SQRT_2 / 2.0, SQRT_2 / 2.0), 0 );
        let i = intersection(SQRT_2, 0);
        let xs = intersections(&[i]);

        let comps = prepare_computations(i, r, &w, &xs);

        assert!{ comps.reflectv.equals( vector(0.0, SQRT_2 / 2.0, SQRT_2 / 2.0) )};
    }

    #[test]
    fn finding_n1_and_n2_at_various_intersections(){
        let mut w = world();

        let mut a = glass_sphere();
        a.set_transform(scaling(2.0, 2.0, 2.0));
        let mut mat = material();
        mat.refractive_index = 1.5;
        a.set_material(mat);
        w.add_object(Box::new(a));

        let mut b = glass_sphere();
        b.set_transform(translation(0.0, 0.0, -0.25));
        let mut mat = material();
        mat.refractive_index = 2.0;
        b.set_material(mat);
        w.add_object(Box::new(b));

        let mut c = glass_sphere();
        c.set_transform(translation(0.0, 0.0, 0.25));
        let mut mat = material();
        mat.refractive_index = 2.5;
        c.set_material(mat);
        w.add_object(Box::new(c));
        
        let r = ray(point(0.0, 0.0, -4.0), vector(0.0, 0.0, 1.0), 0);

        let i1 = intersection(2.00, 0); // more 'correct' to use a.get_index()
        let i2 = intersection(2.75, 1); // here, but that will cause a borrowing
        let i3 = intersection(3.25, 2); // issue and require clone() above, 
        let i4 = intersection(4.75, 1); // simpler to have this index hardcoded
        let i5 = intersection(5.25, 2); // for the purposes of this test
        let i6 = intersection(6.00, 0);
        let xs = intersections(&[i1, i2, i3, i4, i5, i6]);

        let comps = prepare_computations(xs[0], r, &w, &xs);
        assert_eq!(comps.n1, 1.0);
        assert_eq!(comps.n2, 1.5);
        
        let comps = prepare_computations(xs[1], r, &w, &xs);
        assert_eq!(comps.n1, 1.5);
        assert_eq!(comps.n2, 2.0);

        let comps = prepare_computations(xs[2], r, &w, &xs);
        assert_eq!(comps.n1, 2.0);
        assert_eq!(comps.n2, 2.5);

        let comps = prepare_computations(xs[3], r, &w, &xs);
        assert_eq!(comps.n1, 2.5);
        assert_eq!(comps.n2, 2.5);

        let comps = prepare_computations(xs[4], r, &w, &xs);
        assert_eq!(comps.n1, 2.5);
        assert_eq!(comps.n2, 1.5);

        let comps = prepare_computations(xs[5], r, &w, &xs);
        assert_eq!(comps.n1, 1.5);
        assert_eq!(comps.n2, 1.0);
    }

    #[test]
    fn under_point_offset_below_surface(){
        let mut w = world();
        
        let mut shape = glass_sphere();
        shape.set_transform(translation(0.0, 0.0, 1.0));
        w.add_object(Box::new(shape));

        let r = ray(point(0.0, 0.0, -5.0), vector(0.0, 0.0, 1.0), 0);
        let i = intersection(5.0, 0); 
        let xs = intersections(&[i]);

        let comps = prepare_computations(i, r, &w, &xs);

        assert!(comps.under_point.z > EPSILON/2.0);
        assert!(comps.point.z < comps.under_point.z);
    }

    #[test]
    fn schlick_under_total_internal_reflection(){
        let mut w = world();

        let shape = glass_sphere();
        w.add_object(Box::new(shape));

        let r = ray(point(0.0, 0.0, SQRT_2/2.0), vector(0.0, 1.0, 0.0), 0);
        let i1 = intersection(-SQRT_2/2.0, 0); 
        let i2 = intersection( SQRT_2/2.0, 0); 
        let xs = intersections(&[i1, i2]);
        let comps = prepare_computations(xs[1], r, &w, &xs);

        let reflectance = schlick(comps);

        assert!(reflectance == 1.0);
    }

    #[test]
    fn schlick_with_perpendicular_view_angle(){
        let mut w = world();

        let shape = glass_sphere();
        w.add_object(Box::new(shape));

        let r = ray(point(0.0, 0.0, 0.0), vector(0.0, 1.0, 0.0), 0);
        let i1 = intersection(-1.0, 0); 
        let i2 = intersection( 1.0, 0); 
        let xs = intersections(&[i1, i2]);
        let comps = prepare_computations(xs[1], r, &w, &xs);

        let reflectance = schlick(comps);

        assert!(equals(reflectance, 0.04));
    }

    #[test]
    fn schlick_with_small_view_angle(){
        let mut w = world();

        let shape = glass_sphere();
        w.add_object(Box::new(shape));

        let r = ray(point(0.0, 0.99, -2.0), vector(0.0, 0.0, 1.0), 0);
        let i1 = intersection(1.8589, 0); 
        let xs = intersections(&[i1]);
        let comps = prepare_computations(xs[0], r, &w, &xs);

        let reflectance = schlick(comps);

        assert!(equals(reflectance, 0.48873));
    }
}

File: python/intersections.py
# to run tests: python -m unittest -v intersections

import unittest
import math
import spheres
import rays
import tuples
import transformations
import utils
import planes
import triangles

class Intersection:
    def __init__(self, t, obj):
        self.t = t
        self.object = obj

    def __repr__(self):
        return "Intersection(%r, %r)" % (self.t, self.object)

def intersection(t, obj):
    return Intersection(t, obj)

def intersection_with_uv(t, obj, u, v):
    i = Intersection(t, obj)
    i.u = u
    i.v = v
    return i

def intersections(*args):
    result = []
    for arg in args:
        result.append(arg)
    result.sort(key=lambda x: x.t)
    return result

def hit(xs):
    for i in xs:
        if i.t >=0:
            return i

def schlick(comps):
    cos = comps.eyev.dot(comps.normalv)

    if comps.n1 > comps.n2:
        n = comps.n1 / comps.n2
        sin2_t = n ** 2 * (1.0 - cos ** 2)
        if sin2_t > 1.0:
            return 1.0

        cos_t = math.sqrt(1.0 - sin2_t)
        cos = cos_t

    r0 = ((comps.n1 - comps.n2)/(comps.n1 + comps.n2)) ** 2
    return r0 + (1 - r0) * (1 - cos) ** 5

class Computation:
    def __init__(self, intersect, ray):
        self.t = intersect.t
        self.object = intersect.object
        self.point = ray.position(self.t)
        self.eyev = -ray.direction
        self.normalv = self.object.normal_at(self.point)

        if self.normalv.dot(self.eyev) < 0:
            self.inside = True
            self.normalv = -self.normalv
        else:
            self.inside = False

        self.over_point = self.point + self.normalv * utils.EPSILON
        self.under_point = self.point - self.normalv * utils.EPSILON
        self.reflectv = ray.direction.reflect(self.normalv)
        self.n1 = 1.0
        self.n2 = 1.0

def prepare_computations(intersect, ray, xs=None):
    comp = Computation(intersect, ray)

    if xs:
        containers = []
        for i in xs:
            if i == intersect:
                if containers:
                    comp.n1 = containers[-1].material.refractive_index

            if i.object in containers:
                containers.remove(i.object)
            else:
                containers.append(i.object)

            if i == intersect:
                if containers:
                    comp.n2 = containers[-1].material.refractive_index
                break

    return comp

class IntersectionTestCase(unittest.TestCase):
    def test_an_intersection_encapsulates_t_and_object(self):
        s = spheres.sphere()

        i = intersection(3.5, s)

        self.assertEqual(i.t, 3.5)
        self.assertEqual(i.object, s)

    def test_aggregating_intersections(self):
        s = spheres.sphere()
        i1 = intersection(1, s)
        i2 = intersection(2, s)

        xs = intersections(i1, i2)

        self.assertEqual(len(xs), 2)
        self.assertEqual(xs[0].t, 1)
        self.assertEqual(xs[1].t, 2)

    def test_the_hit_when_all_intersections_positive(self):
        s = spheres.sphere()
        i1 = intersection(1, s)
        i2 = intersection(2, s)
        xs = intersections(i2, i1)

        i = hit(xs)

        self.assertEqual(i, i1)
        
    def test_the_hit_when_some_intersections_negative(self):
        s = spheres.sphere()
        i1 = intersection(-1, s)
        i2 = intersection(1, s)
        xs = intersections(i2, i1)

        i = hit(xs)

        self.assertEqual(i, i2)
        
    def test_the_hit_when_all_intersections_negative(self):
        s = spheres.sphere()
        i1 = intersection(-2, s)
        i2 = intersection(-1, s)
        xs = intersections(i2, i1)

        i = hit(xs)

        self.assertEqual(i, None)
        
    def test_hit_is_always_lowest_nonnegative_intersection(self):
        s = spheres.sphere()
        i1 = intersection(5, s)
        i2 = intersection(7, s)
        i3 = intersection(-3, s)
        i4 = intersection(2, s)
        xs = intersections(i1, i2, i3, i4)

        i = hit(xs)

        self.assertEqual(i, i4)

    def test_precomputing_the_state_of_an_intersection(self):
        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))
        s = spheres.sphere()
        i = intersection(4, s)

        comps = prepare_computations(i, r)

        self.assertEqual(comps.t, i.t)
        self.assertEqual(comps.object, i.object)
        self.assertEqual(comps.point, tuples.point(0, 0, -1))
        self.assertEqual(comps.eyev, tuples.vector(0, 0, -1))
        self.assertEqual(comps.normalv, tuples.vector(0, 0, -1))

    def test_hit_when_intersection_occurs_on_outside(self):
        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))
        s = spheres.sphere()
        i = intersection(4, s)

        comps = prepare_computations(i, r)

        self.assertFalse(comps.inside)

    def test_hit_when_intersection_occurs_on_inside(self):
        r = rays.ray(tuples.point(0, 0, 0), tuples.vector(0, 0, 1))
        s = spheres.sphere()
        i = intersection(1, s)

        comps = prepare_computations(i, r)

        self.assertEqual(comps.point, tuples.point(0, 0, 1))
        self.assertEqual(comps.eyev, tuples.vector(0, 0, -1))
        self.assertTrue(comps.inside)
        self.assertEqual(comps.normalv, tuples.vector(0, 0, -1))

    def test_hit_should_offset_point(self):
        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))
        s = spheres.sphere()
        s.transform = transformations.translation(0, 0, 1)
        i = intersection(5, s)

        comps = prepare_computations(i, r)

        self.assertTrue(comps.over_point.z < -utils.EPSILON/2)
        self.assertTrue(comps.point.z > comps.over_point.z)

    def test_precomputing_reflection_vector(self):
        shape = planes.plane()
        r = rays.ray(tuples.point(0, 1, -1), tuples.vector(0, -math.sqrt(2)/2, math.sqrt(2)/2))
        i = intersection(math.sqrt(2), shape)

        comps = prepare_computations(i, r)

        self.assertEqual(comps.reflectv, tuples.vector(0, math.sqrt(2)/2, math.sqrt(2)/2))

    def test_finding_n1_and_n2_at_various_intersections(self):
        a = spheres.glass_sphere()
        a.set_transform(transformations.scaling(2, 2, 2))
        a.material.refractive_index = 1.5

        b = spheres.glass_sphere()
        b.set_transform(transformations.translation(0, 0, -0.25))
        b.material.refractive_index = 2.0

        c = spheres.glass_sphere()
        c.set_transform(transformations.translation(0, 0, 0.25))
        c.material.refractive_index = 2.5

        r = rays.ray(tuples.point(0, 0, -4), tuples.vector(0, 0, 1))
        xs = intersections(intersection(2, a),
                           intersection(2.75, b),
                           intersection(3.25, c),
                           intersection(4.75, b),
                           intersection(5.25, c),
                           intersection(6, a))
        
        comps = prepare_computations(xs[0], r, xs)
        self.assertEqual(comps.n1, 1.0)
        self.assertEqual(comps.n2, 1.5)
        
        comps = prepare_computations(xs[1], r, xs)
        self.assertEqual(comps.n1, 1.5)
        self.assertEqual(comps.n2, 2.0)

        comps = prepare_computations(xs[2], r, xs)
        self.assertEqual(comps.n1, 2.0)
        self.assertEqual(comps.n2, 2.5)

        comps = prepare_computations(xs[3], r, xs)
        self.assertEqual(comps.n1, 2.5)
        self.assertEqual(comps.n2, 2.5)

        comps = prepare_computations(xs[4], r, xs)
        self.assertEqual(comps.n1, 2.5)
        self.assertEqual(comps.n2, 1.5)

        comps = prepare_computations(xs[5], r, xs)
        self.assertEqual(comps.n1, 1.5)
        self.assertEqual(comps.n2, 1.0)

    def test_under_point_is_offset_below_surface(self):
        shape = spheres.glass_sphere()
        shape.set_transform(transformations.translation(0, 0, 1))

        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))
        xs = intersections(intersection(5, shape))
        comps = prepare_computations(xs[0], r, xs)

        self.assertTrue(comps.under_point.z > utils.EPSILON/2)
        self.assertTrue(comps.point.z < comps.under_point.z)

    def test_schlick_approximation_under_total_internal_reflection(self):
        shape = spheres.glass_sphere()

        r = rays.ray(tuples.point(0, 0, math.sqrt(2)/2), tuples.vector(0, 1, 0))
        xs = intersections(intersection(-math.sqrt(2)/2, shape),
                           intersection( math.sqrt(2)/2, shape))
        comps = prepare_computations(xs[1], r, xs)

        reflectance = schlick(comps)

        self.assertEqual(reflectance, 1.0)

    def test_schlick_approximation_with_perpendicular_viewing_angle(self):
        shape = spheres.glass_sphere()

        r = rays.ray(tuples.point(0, 0, 0), tuples.vector(0, 1, 0))
        xs = intersections(intersection(-1, shape),
                           intersection( 1, shape))
        comps = prepare_computations(xs[1], r, xs)

        reflectance = schlick(comps)

        self.assertTrue(utils.flequal(reflectance, 0.04))

    def test_schlick_approximation_with_small_angle_and_n2_gt_n1(self):
        shape = spheres.glass_sphere()

        r = rays.ray(tuples.point(0, 0.99, -2), tuples.vector(0, 0, 1))
        xs = intersections(intersection(1.8589, shape))
        comps = prepare_computations(xs[0], r, xs)

        reflectance = schlick(comps)

        self.assertTrue(utils.flequal(reflectance, 0.48874))

    def test_an_intersection_can_encapsulate_u_and_v(self):
        s = triangles.triangle(tuples.point( 0, 1, 0),
                               tuples.point(-1, 0, 0),
                               tuples.point( 1, 0, 0))

        i = intersection_with_uv(3.5, s, 0.2, 0.4)

        self.assertEqual(i.u, 0.2)
        self.assertEqual(i.v, 0.4)

# ---------------------------------------------------------------------------
if __name__ == '__main__':
    unittest.main()
